File: 2024/march/29 Count Subarrays Where Max Element Appears Atleast K Times.cpp
// 2962. Count Subarrays Where Max Element Appears At Least K Times
// https://leetcode.com/problems/count-subarrays-where-max-element-appears-at-least-k-times/

class Solution {
public:
    long long countSubarrays(vector<int>& nums, int k) {
        int maxElement = *max_element(nums.begin(), nums.end());
        vector<int> indexesOfMaxElement;
        long long answer = 0;

        for (int i = 0; i < nums.size(); i++) {
            if (nums[i] == maxElement) {
                indexesOfMaxElement.push_back(i);
            }

            int freq = indexesOfMaxElement.size();
            if (freq >= k) {
                answer += indexesOfMaxElement[freq - k] + 1;
            }
        }

        return answer;
    }
};

File: 2025/04 april/29 count-subarrays-where-max-element-appears-at-least-k-times.rs
// 2962. count-subarrays-where-max-element-appears-at-least-k-times
// https://leetcode.com/problems/count-subarrays-where-max-element-appears-at-least-k-times

impl Solution {
    pub fn count_subarrays(nums: Vec<i32>, k: i32) -> i64 {
        let max = *nums.iter().max().unwrap();
        let (mut left, mut count, mut res) = (0, 0, 0 as i64);

        for val in nums.iter() {
            if *val == max {
                count += 1;
            }

            while count >= k {
                if (nums[left] == max) {
                    count -= 1;
                }
                left += 1;
            }

            res += left as i64;
        }
        res
    }
}
